[DI-27060] - Alerts Contextual View - Api flexibility

ankita-akamai opened the pull request:
## Description 📝

Highlight the Pull Request's context and intentions.

## Changes  🔄

List any change(s) relevant to the reviewer.

- ...
- ...

### Scope 🚢

 Upon production release, changes in this PR will be visible to:

- [ ] All customers
- [ ] Some customers (e.g. in Beta or Limited Availability)
- [ ] No customers / Not applicable

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷

**Include a screenshot `<img src="" />` or video `<video src="" />` of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: For changes requiring multiple steps to validate, prefer a video for clarity.

| Before  | After   |
| ------- | ------- |
| 📷 | 📷 |

## How to test 🧪

### Prerequisites

(How to setup test environment)

- ...
- ...

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] ...
- [ ] ...

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

*Check all that apply*

- [ ] Use React components instead of HTML Tags
- [ ] Proper naming conventions like cameCase for variables & Function & snake_case for constants
- [ ] Use appropriate types & avoid using "any"
- [ ] No type casting & non-null assertions
- [ ] Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] Providing/Improving test coverage
- [ ] Use sx props to pass styles instead of style prop
- [ ] Add JSDoc comments for interface properties & functions
- [ ] Use strict equality (===) instead of double equal (==)
- [ ] Use of named arguments (interfaces) if function argument list exceeds size 2
- [ ] Destructure the props
- [ ] Keep component size small & move big computing functions to separate utility
- [ ] 📱 Providing mobile support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All tests and CI checks are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

<!-- This content will not appear in the rendered Markdown 

## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**

- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

-->

ankita-akamai commented on packages/manager/src/features/Linodes/LinodesDetail/LinodeAlerts/LinodeAlerts.tsx:
```diff
@@ -32,27 +32,21 @@ const LinodeAlerts = () => {
     <Box>
-      {aclpServices?.linode?.alerts?.enabled &&
-        isAclpAlertsSupportedRegionLinode && (
+      {(
```
`LinodeAlerts.tsx` are temporary changes, will be removed

ankita-akamai commented on packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts:
```diff
@@ -629,3 +636,68 @@ export const transformDimensionValue = (
+  const alertIdsToInvalidate = Array.from(
+    new Set([...oldEnabledAlertIds, ...newEnabledAlertIds])
+  );
+
```
Just noticed that it is better to update cache on onsucces to handle the lag after PUT call response, exploring!!

ankita-akamai commented:
**`[Note]`**: Test cases will be pushed once current generic mutation wrapper is finalized.

venkymano-akamai commented:
@ankita-akamai , Seems we need _KT session for this_!! , I will review it, this is complex as well, lets see how it differs from a basic onboarding and advantages of it.

venkymano-akamai commented:
```
export const useServiceAlertsMutation = (
  serviceType: string,
  entityId: string
) => {
  const queryClient = useQueryClient();
  return useMutation<{}, APIError[], CloudPulseAlertsPayload>({
    mutationFn: (payload: CloudPulseAlertsPayload) => {
      if (serviceType === 'linode') {
        // Handle Linode specific logic
        return updateLinode(Number(entityId), getLinodeAlertsPayload(payload));
      } else if (serviceType === 'dbaas') {
        return updateDatabase(
          'mysql',
          Number(entityId),
          getDatabaseAlertsPayload(payload)
        );
      }
      return updateServiceAlerts(serviceType, entityId, payload);
    },
    onSuccess(_, payload) {
      const allAlerts = queryClient.getQueryData<Alert[]>(
        queryFactory.alerts._ctx.all().queryKey
      );

      // Get alerts previously enabled for this entity
      const oldEnabledAlertIds =
        allAlerts
          ?.filter((alert) => alert.entity_ids.includes(entityId))
          .map((alert) => [alert.id](https://alert.id/)) || [];

      // Combine enabled user and system alert IDs from payload
      const newEnabledAlertIds = [
        ...(payload.user ?? []),
        ...(payload.system ?? []),
      ];

      // Get unique list of all enabled alert IDs for cache invalidation
      const alertIdsToInvalidate = Array.from(
        new Set([...oldEnabledAlertIds, ...newEnabledAlertIds])
      );

      queryClient.invalidateQueries({
        queryKey: queryFactory.resources(serviceType).queryKey,
      });

      queryClient.invalidateQueries({
        queryKey: queryFactory.alerts._ctx.all().queryKey,
      });

      alertIdsToInvalidate.forEach((alertId) => {
        queryClient.invalidateQueries({
          queryKey: queryFactory.alerts._ctx.alertByServiceTypeAndId(
            serviceType,
            String(alertId)
          ).queryKey,
        });
      });
    },
  });
};

const getLinodeAlertsPayload = (
  payload: CloudPulseAlertsPayload
): DeepPartial<Linode> => {
  return {
    alerts: {
      user: payload.user,
      system: payload.system,
    },
  };
};

const getDatabaseAlertsPayload = (
  payload: CloudPulseAlertsPayload
): UpdateDatabasePayload => {
  return {
    user: payload.user,
    system: payload.system,
  };
};
```

Thinking of something like this

venkymano-akamai commented:
Sorry I accidently closed and opened

ankita-akamai commented:
@venkymano-akamai  @nikhagra-akamai  I removed all typecasting, brought all usehooks to a common hook, kept the payload transformation as it is. Added test case for one util. Please review again.

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts:
```diff
@@ -626,3 +633,68 @@ export const transformDimensionValue = (
+  });
+};
+
+/**
+ * Checks if two arrays are equal, ignores the order of the elements
+ * @param a The first array
+ * @param b The second array
+ * @returns True if the arrays are equal, false otherwise
+ */
+export const arraysEqual = (a: number[] | undefined, b: number[] | undefined) =>
+  compareArrays(
+    [...(a ?? [])].sort((x, y) => x - y),
+    [...(b ?? [])].sort((x, y) => x - y)
+  );
```
@ankita-akamai , there is a common util for the entire code base, check if something similar is there?

nikhagra-akamai replied:
```suggestion

if(a=== undefined && b === undefined ) return true;
if(a === undefined || b === undefined ) return false;

if(a.length !== b.length) return false;
  compareArrays(
    [...(a ?? [])].sort((x, y) => x - y),
    [...(b ?? [])].sort((x, y) => x - y)
  );
```

put more checks before sorting it

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/ContextualView/AlertInformationActionTable.tsx:
```diff
@@ -151,10 +155,8 @@ export const AlertInformationActionTable = (
-    entityId ?? ''
-  );
+  // Mutation to update alerts as per service type
+  const updateAlerts = useAlertsMutation(serviceType, entityId ?? '');
```
```suggestion
  const updateAlerts = useAlertsMutation(serviceType, entityId);
```

lets not use any fallback value instead handle undefined separately

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/ContextualView/AlertInformationActionTable.tsx:
```diff
@@ -167,19 +169,26 @@ export const AlertInformationActionTable = (
     setIsDialogOpen(false);
   };
 
+  const qc = useQueryClient();
```
```suggestion
  const queryClient = useQueryClient();
```

nikhagra-akamai commented on packages/manager/src/queries/cloudpulse/useAlertsMutation.ts:
```diff
@@ -0,0 +1,72 @@
+ * @param entityId entity id
+ * @returns alerts mutation
+ */
+export const useAlertsMutation = <T extends CloudPulseServiceType>(
```
Why generic? just put CloudPulseServiceType

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/ContextualView/AlertInformationActionTable.tsx:
```diff
@@ -167,19 +169,26 @@ export const AlertInformationActionTable = (
+          invalidateAlerts(qc, serviceType, entityId ?? '', payload);
         })
         .catch(() => {
           enqueueSnackbar('Alerts changes were not saved, please try again.', {
```
keep all these logics inside hook function only. Also move invalidate Alert function from utils to mutation file because it is more specific for that

nikhagra-akamai commented on packages/manager/src/queries/cloudpulse/useAlertsMutation.ts:
```diff
@@ -0,0 +1,72 @@
+      return updateLinode;
+    default:
+      return updateServiceAlerts;
+  }
```
```suggestion
  const updateAlerts = (payload: CloudPulseAlertPayload) => {
  
  switch (serviceType) {
    case 'linode':
      //handle function calls here and add call back handler for success & failure handling
    default:
      //handle function calls here
  }
  
  }
  
  return updaetAlerts
```

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts:
```diff
@@ -626,3 +633,68 @@ export const transformDimensionValue = (
+export const invalidateAlerts = (
+  qc: QueryClient,
+  serviceType: string,
+  entityId: string,
```
here as well handle entityId as undefined instead of empty fallback value

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/Utils/utils.ts:
```diff
@@ -626,3 +633,68 @@ export const transformDimensionValue = (
+  // Get unique list of all enabled alert IDs for cache invalidation
+  const alertIdsToInvalidate = Array.from(
+    new Set([...oldEnabledAlertIds, ...newEnabledAlertIds])
+  );
```
```suggestion
  const alertIdsToInvalidate = [...oldEnabledAlertIds, ...newEnabledAlertIds];
```
in worst case it will execute 2*size of alerts times, its okay it'll not make much difference
